src/main/scala/com/microsoft/hyperspace/index/dataskipping/util/extractors.scala

```diff
@@ -0,0 +1,199 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.dataskipping.util
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.types.{BooleanType, DataType}
+
+// Value extractors returning Option[Literal] check if the value is not null
+// because we're only interested in non-null values.
+//
+// Also note that we don't go overboard to match every pattern because
+// we assume that Catalyst optimizer will give us an optimized predicate in NNF.
+// It means in general we don't have to deal with Not, or worry about
+// foldable expressions because they will be optimized to literals.
+//
+// There are some differences between Spark versions and that's why we include
+// some patterns that are no longer needed in newer Spark versions.
+
+/**
+ * Extracts the non-null literal value in the predicate if it's equivalent to
+ * <expr> = <literal>.
+ *
+ * For the purpose of data skipping, we don't extract the value if it's null.
+ * If the literal is not null, then the only way to make the predicate
+ * <expr> = <literal> or <expr> <=> <literal> is when the expression is not null
+ * and its value is equal to the literal value.
+ */
+case class EqualToExtractor(matcher: ExprMatcher) {
+  def unapply(p: Expression): Option[Literal] =
+    p match {
+      case EqualTo(e, v: Literal) if v.value != null && matcher(e) => Some(v)
+      case EqualTo(v: Literal, e) if v.value != null && matcher(e) => Some(v)
+      case EqualNullSafe(e, v: Literal) if v.value != null && matcher(e) => Some(v)
+      case EqualNullSafe(v: Literal, e) if v.value != null && matcher(e) => Some(v)
+      case _ => None
+    }
+}
+
+/**
+ * Extracts the non-null literal value in the predicate if it's equivalent to
+ * <expr> < <literal>.
+ */
+case class LessThanExtractor(matcher: ExprMatcher) {
+  def unapply(p: Expression): Option[Literal] =
+    p match {
+      case LessThan(e, v: Literal) if v.value != null && matcher(e) => Some(v)
+      case GreaterThan(v: Literal, e) if v.value != null && matcher(e) => Some(v)
+      case _ => None
+    }
+}
+
+/**
+ * Extracts the non-null literal value in the predicate if it's equivalent to
+ * <expr> <= <literal>.
+ */
+case class LessThanOrEqualToExtractor(matcher: ExprMatcher) {
+  def unapply(p: Expression): Option[Literal] =
+    p match {
+      case LessThanOrEqual(e, v: Literal) if v.value != null && matcher(e) => Some(v)
+      case GreaterThanOrEqual(v: Literal, e) if v.value != null && matcher(e) => Some(v)
+      case _ => None
+    }
+}
+
+/**
+ * Extracts the non-null literal value in the predicate if it's equivalent to
+ * <expr> > <literal>.
+ */
+case class GreaterThanExtractor(matcher: ExprMatcher) {
```

Comment:
Do we need to define both GreaterThanExtractor & LessThanExtractor? 
I think one is enough but if there's some reason then it's fine

Reply:
We need both. GreaterThanExtractor extracts patterns like `A > 1` and `1 < A`. LessThanExtractor extracts patterns like `A < 1` and `1 > A`.